src/modelfree/transparent.py

```diff
@@ -0,0 +1,188 @@
+    - If key is present, value (bool) corresponds to whether we augment the observation space
+    with it. This is because TransparentCurryVecEnv needs this information to modify its
+    observation space, and we want all of the transparency-related parameters in one dict.
+    """
+    def __init__(self, transparent_params):
+        if transparent_params is None:
+            raise ValueError("TransparentPolicy requires transparent_params.")
+        for key in transparent_params:
+            if key not in TRANSPARENCY_KEYS:
+                raise KeyError(f"Unrecognized transparency key: {key}")
+        self.transparent_params = transparent_params
+
+    @abstractmethod
+    def get_obs_sizes(self):
+        """List of dimensionalities of different data that could be appended to an observation.
+        Specifically (observations, (concatenated) policy acts., value acts., and hidden state)
+
+        This is used in get_obs_aug_amount, which has the common logic for all subclasses.
+        If class does not support a key, its value should be None.
+
+        :return: (list<int,None>) sizes of data from TRANSPARENCY_KEYS for this policy.
+        """
+        raise NotImplementedError()
+
+    def get_obs_aug_amount(self):
+        """Execute loop for calculating amount by which to augment observations"""
+        obs_aug_amount = 0
+        obs_sizes = self.get_obs_sizes()
+        for key, val in list(zip(TRANSPARENCY_KEYS, obs_sizes)):
+            if self.transparent_params.get(key):
+                obs_aug_amount += val
+        return obs_aug_amount
+
+    def get_transparency_dict(self, transparent_objs):
+        """Construct dictionary of transparent data
+
+        :param transparent_objs: (tuple<np.ndarray,None>) values for keys in TRANSPARENCY_KEYS
+        :return: (dict[str, np.ndarray]) dict with all exposed data
+        """
+        transparency_dict = {k: np.squeeze(v) for k, v in zip(TRANSPARENCY_KEYS, transparent_objs)
+                             if k in self.transparent_params}
+        return transparency_dict
+
+
+class TransparentFeedForwardPolicy(TransparentPolicy, FeedForwardPolicy):
+    """stable_baselines FeedForwardPolicy which is also transparent."""
+    def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, transparent_params,
+                 reuse=False, layers=None, net_arch=None, act_fun=tf.tanh,
+                 cnn_extractor=nature_cnn, feature_extraction="cnn", **kwargs):
+        FeedForwardPolicy.__init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse,
+                                   layers, net_arch, act_fun, cnn_extractor, feature_extraction,
+                                   **kwargs)
+        TransparentPolicy.__init__(self, transparent_params)
+
+    def get_obs_sizes(self):
+        return self.ob_space.shape[0], sum(self.layers), sum(self.layers), None
+
+    def step(self, obs, state=None, mask=None, deterministic=False):
```

Comment:
The code duplication here with `FeedForwardPolicy` is a bit upsetting. The simple approach would be to call the base-class `step`, and then make another call to `self.sess.run([self.ff_out], {self.obs_ph: obs})`. You could even have a default implementation in `TransparentPolicy` that does something like this: it'd work for a broad range of policies.

However, this would incur some performance hit as we'd be calling TensorFlow twice. Maybe worth benchmarking to check this is actually significant? @decodyng, any opinion on how bad the code duplication is currently?

Reply:
It definitely is frustrating, but most of the things I'm coming up with as alternatives to the current setup or the proposed "call it twice" approach would require even more aggressive monkey-patching of stable_baselines itself to allow for more flexible sess.run() call patterns, which seems arguably worse. I'd agree with you, @AdamGleave, in having a preference for running .step() and then calling a separate sess.run() on the transparent params, at least for some range of values for how costly that duplication turns out to be; there's obviously a point where I think the duplication is worth avoiding the performance hit.